File: projects/2026-09-22-kmpfullstack-short-link-service/iosApp/Sources/ShortLinkApp/Data/LinkRepository.swift
import Foundation

/// The client-side contract — same shape as the `LinkRepository` interface in
/// `shared/src/commonMain`, reimplemented in Swift since Kotlin/Native can't cross-compile
/// an iOS `.xcframework` on this Linux runner (see the project README for details).
protocol LinkRepository {
    func list() async throws -> [ShortLink]
    func create(longUrl: String) async throws -> ShortLink
    func visit(code: String) async throws -> ShortLink?
}

enum LinkRepositoryError: LocalizedError {
    case invalidUrl(String)
    case server(String)
    case notFound(String)

    var errorDescription: String? {
        switch self {
        case .invalidUrl(let message), .server(let message), .notFound(let message):
            return message
        }
    }
}

File: projects/2026-09-22-kmpfullstack-short-link-service/shared/src/commonMain/kotlin/com/dailyprojects/shortlink/LinkRepository.kt
package com.dailyprojects.shortlink

/** The one contract every target codes against — an in-memory store on the server, HTTP on every client. */
interface LinkRepository {
    suspend fun list(): List<ShortLink>
    suspend fun create(longUrl: String): ShortLink
    suspend fun visit(code: String): ShortLink?
}
